File: packages/contracts/src/standards.ts
/** CrowdFlow's constants registry. Every assumption is named and documented. */

export const LOS_A_MAX = 23;
export const LOS_B_MAX = 33;
export const LOS_C_MAX = 49;
export const LOS_D_MAX = 66;
export const LOS_E_MAX = 82;

export const FREE_FLOW_SPEED_MS = 1.34;
export const JAM_DENSITY_PERSONS_M2 = 4;
export const OBSERVED_HIGH_DENSITY_PERSONS_M2 = 4.7;
export const CAPACITY_DENSITY = JAM_DENSITY_PERSONS_M2 / 2;

import type { LOSBand } from './types.js';

export function densityForFlow(
  flowPedMMin: number,
  freeSpeedMs = FREE_FLOW_SPEED_MS,
  jamDensity = JAM_DENSITY_PERSONS_M2,
): number | null {
  const c = (flowPedMMin * jamDensity) / (60 * freeSpeedMs);
  const discriminant = jamDensity ** 2 - 4 * c;
  return discriminant < 0 ? null : (jamDensity - Math.sqrt(discriminant)) / 2;
}

export const DENSITY_NOMINAL_MAX = densityForFlow(LOS_C_MAX) ?? 0.75;
export const DENSITY_BUILDING_MAX = CAPACITY_DENSITY;

/** Authoritative classifier. Flow is non-monotonic and must never classify a zone. */
export function bandForDensity(personsM2: number): LOSBand {
  if (personsM2 < DENSITY_NOMINAL_MAX) return 'nominal';
  if (personsM2 < DENSITY_BUILDING_MAX) return 'building';
  return 'critical';
}

export function losGradeForFlow(flowPedMMin: number): string {
  for (const [grade, max] of [
    ['A', LOS_A_MAX],
    ['B', LOS_B_MAX],
    ['C', LOS_C_MAX],
    ['D', LOS_D_MAX],
    ['E', LOS_E_MAX],
  ] as const) {
    if (flowPedMMin < max) return grade;
  }
  return 'F';
}

export const ASSUMED_REPORTABLE_CONFIDENCE_FLOOR = 0.25;
export const ASSUMED_REPORTABLE_NODE_FLOOR = 3;
export const ASSUMED_ACTIONABLE_PROBABILITY_FLOOR = 0.6;
export const ASSUMED_ACTIONABLE_CONFIDENCE_FLOOR = 0.5;
export const ASSUMED_CONFIDENCE_COUNT_SATURATION = 200;
export const ASSUMED_CONFIDENCE_COUNT_WEIGHT = 0.55;
export const ASSUMED_POSITION_ACCURACY_BEST_M = 5;
export const ASSUMED_POSITION_ACCURACY_WORST_M = 50;
export const ASSUMED_ROUTE_CACHE_ENTRIES = 4096;
export const ASSUMED_ORPHAN_ZONE_LENGTH_M = 25;
export const ASSUMED_ORPHAN_ZONE_WIDTH_M = 2;

export const MEASURED_SAMPLE_FLOOR = 30;
export const MAD_TO_SIGMA = 1.4826;
export const MODIFIED_Z_OUTLIER = 3.5;

export const ASSUMED_RADIO_RANGE_CROWD_M = 30;
export const PROPHET_P_INIT = 0.75;
export const PROPHET_BETA = 0.25;
export const PROPHET_GAMMA = 0.98;
export const ASSUMED_SPRAY_COPY_SCALING = 1;
export const MESH_TTL_MAX = 8;
export const ASSUMED_HOP_LATENCY_S = 5;
export const ASSUMED_MESH_BUFFER_MESSAGES = 256;
export const ASSUMED_URGENT_RELAYS_PER_MIN = 30;
export const ASSUMED_URGENT_BURST_RELAYS = 15;
export const ASSUMED_UPLINK_BATTERY_RESERVE = 0.2;
export const ASSUMED_SKEW_WINDOW_S = 300;
export const ASSUMED_DEMO_POPULATION = 2500;
export const ASSUMED_ANTHROPIC_THINKING_BUDGET_TOKENS = 4000;

export function prophetTimeUnitS(
  radioRangeM = ASSUMED_RADIO_RANGE_CROWD_M,
  walkSpeedMs = FREE_FLOW_SPEED_MS,
  gamma = PROPHET_GAMMA,
): number {
  return (radioRangeM / walkSpeedMs) * Math.log(1 / gamma) / Math.log(2);
}

export function sprayCopiesFor(reachableNodes: number): number {
  if (reachableNodes <= 0) return 2;
  return Math.max(2, Math.ceil(ASSUMED_SPRAY_COPY_SCALING * Math.sqrt(reachableNodes)));
}

export function dedupeRetentionS(
  ttl = MESH_TTL_MAX,
  hopLatencyS = ASSUMED_HOP_LATENCY_S,
): number {
  return ttl * hopLatencyS;
}

export const GEOIND_PRIVACY_LEVEL = Math.log(4);
export const ASSUMED_GEOIND_RADIUS_M = 50;
export const GEOIND_EPSILON_VENUE = GEOIND_PRIVACY_LEVEL / ASSUMED_GEOIND_RADIUS_M;
export const ASSUMED_FRAGMENT_MAX_DURATION_S = 120;
export const CAPTURE_RECAPTURE_MIN_SAMPLE = 2;
export const CAPTURE_RECAPTURE_MIN_OVERLAP = 1;
export const ASSUMED_PRIVATE_SKETCH_K = 32;
export const ASSUMED_PRIVATE_SKETCH_EPSILON = 1;

/**
 * Radio positioning.
 *
 * Every number here is a starting value for a curve that must be walked before
 * it can be trusted. `rssi_at_1m` and the path-loss exponent are the intercept
 * and the slope of the same line, and the line is a property of one
 * installation in one crowd — an anchor behind a metal panel and one on a mast
 * share a datasheet and nothing else. So these are ASSUMED, they are the
 * defaults a `RadioAnchor` overrides with its own Sourced values, and
 * 'path_loss_exponent' is on the measured-not-assumed list below.
 */

/** Free-space reference for a 2.4 GHz AP at one metre. */
export const ASSUMED_WIFI_RSSI_AT_1M_DBM = -40;
/** The iBeacon convention, which is a one-metre reference and not a 2.4 GHz law. */
export const ASSUMED_BLE_RSSI_AT_1M_DBM = -59;
/** Free space is 2.0. Open paddock tarmac with line of sight sits just above it. */
export const ASSUMED_PATH_LOSS_EXPONENT_OPEN = 2.2;
/**
 * A packed concourse. Water absorbs 2.4 GHz and a crowd is mostly water, so the
 * exponent rises with the density it is measuring — which is why a fix degrades
 * exactly where the system most needs one.
 */
export const ASSUMED_PATH_LOSS_EXPONENT_CROWD = 3.3;
/** Under a grandstand: concrete, steel and no line of sight to anything. */
export const ASSUMED_PATH_LOSS_EXPONENT_COVERED = 3.8;

/** Below three, the geometry is not a fix. Two anchors give a weighted centroid, and the contract says so via PositionFix.anchors_used. */
export const ASSUMED_MIN_ANCHORS_FOR_FIX = 3;
/** An observation older than this describes where the phone was, not where it is. */
export const ASSUMED_ANCHOR_OBSERVATION_TTL_S = 20;
/**
 * Android throttles foreground Wi-Fi scans to four per two minutes since 9
 * (Pie). A platform fact, not a preference: ask for more and the extra calls
 * return the previous results with old timestamps, which is worse than not
 * asking. Everything about the Wi-Fi cadence follows from this one number.
 */
export const ANDROID_WIFI_SCAN_THROTTLE_PER_2_MIN = 4;
export const ASSUMED_WIFI_SCAN_INTERVAL_S = 120 / ANDROID_WIFI_SCAN_THROTTLE_PER_2_MIN;
/** A BLE scan window long enough to hear a beacon advertising at 1 Hz more than once. */
export const ASSUMED_BLE_SCAN_WINDOW_S = 4;
/** GNSS is cheap to sample and the fused provider caches, so this is the loop cadence rather than a duty cycle. */
export const ASSUMED_GNSS_SAMPLE_INTERVAL_S = 10;

/**
 * A fix wider than this is not a position, it is a zone name with extra steps.
 * Rejected outright rather than reported with a large sigma, because the state
 * engine would place it in a zone and the operator would count it there.
 */
export const ASSUMED_FIX_ACCURACY_CEILING_M = 60;
/** No radio solve can beat this; anything tighter is the solver flattering itself. */
export const ASSUMED_FIX_ACCURACY_FLOOR_M = 3;
/** After this, the last fix is history and the node stops reporting rather than repeating itself. */
export const ASSUMED_FIX_STALE_S = 45;
/**
 * How much better a challenger must be before the fuser switches source.
 * Without hysteresis two sources of similar quality trade the lead every tick,
 * and the resulting position jitters between two solutions that are each fine —
 * which reads on a console as a person vibrating.
 */
export const ASSUMED_SOURCE_SWITCH_HYSTERESIS = 1.25;
/**
 * How long a moving node may be carried on its last velocity with no new fix.
 * Long enough to cross a Wi-Fi scan gap, short enough that it cannot walk a
 * phone through a wall and into the next zone.
 */
export const ASSUMED_DEAD_RECKONING_MAX_S = 15;
/** Below this, reported speed is sensor noise and heading is meaningless. */
export const ASSUMED_HEADING_SPEED_FLOOR_MS = 0.3;
/** Displacement smoothing across fixes: fully responsive is jitter, fully damped is a stale dot. */
export const ASSUMED_VELOCITY_SMOOTHING = 0.4;

/** How often the handset's pseudonym rotates. An id that outlives the walk is a trail. */
export const ASSUMED_ID_ROTATION_S = 900;
/** Samples per upload batch. Bounded so an uplink that has been unreachable for an hour cannot arrive as one enormous request. */
export const ASSUMED_UPLINK_BATCH_MAX = 60;
export const ASSUMED_UPLINK_INTERVAL_S = 30;
/** Retained sensing beyond this is deleted, reported or not. The disclosure says 24 hours; this is that sentence in code. */
export const RETENTION_MAX_S = 24 * 60 * 60;

/**
 * The disclosure a report cites.
 *
 * Versioned because the sentence people agreed to is part of the data. If the
 * wording of what the app promises changes, reports gathered under the old
 * wording are not retroactively covered by the new one — so the version travels
 * with every batch and the server rejects an id it does not serve, rather than
 * accepting the data and sorting the consent question out later.
 */
export const LOCATION_DISCLOSURE_VERSION = 'location-disclosure.v1';

/** Disclosure versions still honoured. Withdraw one by removing it here: every
 *  handset still citing it stops being accepted, and stops sensing, on the next
 *  batch — because `IngestAck.stop` says so. */
export const SERVED_DISCLOSURE_VERSIONS = [LOCATION_DISCLOSURE_VERSION] as const;

/**
 * How long a session lasts when nobody published an end time.
 *
 * Jolpica gives session START times only. An end has to come from somewhere, and
 * these are the sporting regulations' scheduled durations — a practice hour, a
 * qualifying hour, a race capped at two. They are ASSUMED because a session that
 * is red-flagged, shortened for weather or run to the two-hour limit ends when it
 * ends, and the chequered flag is the single largest crowd-movement trigger of
 * the day. Every session built from these carries `end_provenance: 'assumed'` so
 * nothing downstream mistakes the number for a schedule.
 */
export const ASSUMED_SESSION_MINUTES: Record<string, number> = {
  practice: 60,
  qualifying: 60,
  sprint: 30,
  race: 120,
  support: 45,
  ceremony: 30,
};

export const MEASURED_NOT_ASSUMED = [
  'participation_rate',
  'zone_capacity',
  'corridor_width',
  'walking_speed',
  'time_to_congestion',
  'prediction_confidence',
  'radio_range',
  'hop_latency',
  'uplink_coverage',
  'path_loss_exponent',
  'anchor_position',
  'fix_accuracy',
] as const;

File: packages/contracts/src/crowdflow_contracts/standards.py
"""Published constants, each with its source.

No magic numbers. Every threshold here cites a standard; anything that cannot be
cited must be measured at runtime instead. See plan/standards.md for the registry
and the reasoning.

The governing framework is Fruin's Level of Service for walkways, expressed as a
flow rate in pedestrians per metre of width per minute. It is the basis of most
crowd-safety practice, and its boundaries are the ones safety practitioners use --
which is why the system's alert thresholds are Fruin's numbers rather than ours.
"""

from __future__ import annotations

from enum import Enum

# --------------------------------------------------------------------------
# Fruin Level of Service, walkways.
# Source: Fruin, "Pedestrian Planning and Design" (1971).
# Boundaries via https://docs.idigitaltwin.org/docs/peddesign/service-level-analysis/
# Units: pedestrians per metre of width per minute.
# --------------------------------------------------------------------------

LOS_A_MAX = 23.0
"""Free flow; slower pedestrians can be bypassed."""

LOS_B_MAX = 33.0
"""Normal walking speed; bypassing possible in one-way flow."""

LOS_C_MAX = 49.0
"""Some speeds restricted; two-way flow needs frequent adjustment."""

LOS_D_MAX = 66.0
"""Majority of walking speeds restricted; frequent conflicts."""

LOS_E_MAX = 82.0
"""Frequent stoppages and interruptions to flow. Above this, flow breaks down.

Commonly cited as the UK Green Guide flow figure for level surfaces, which would
make this the same number the safety regulator uses. That correspondence is
UNVERIFIED against the current edition -- do not claim it publicly until someone
has read it. See plan/standards.md section 2.
"""

# --------------------------------------------------------------------------
# The three operational bands.
#
# Operators act on three states, so the six LOS grades collapse to three -- but
# on Fruin's boundaries, not ones we chose. The middle band is the entire
# product: below BUILDING there is nothing to do, above it it is already too late.
# --------------------------------------------------------------------------

BAND_NOMINAL_MAX = LOS_C_MAX   # 49.0 -- people walk at their chosen speed
BAND_BUILDING_MAX = LOS_E_MAX  # 82.0 -- restricted but flowing: the intervention window


class LOSBand(str, Enum):
    """Operational density band. Always rendered with its word and its number."""

    NOMINAL = "nominal"
    BUILDING = "building"
    CRITICAL = "critical"

    @property
    def label(self) -> str:
        return self.value.upper()

    @property
    def los_grades(self) -> str:
        return {"nominal": "A-C", "building": "D-E", "critical": "F"}[self.value]


def band_for_flow(flow_ped_m_min: float) -> LOSBand:
    """Classify a flow rate into an operational band.

    Args:
        flow_ped_m_min: pedestrians per metre of width per minute.

    The only place these boundaries are applied. If a band appears anywhere else
    in the codebase, it is a bug.
    """
    if flow_ped_m_min < BAND_NOMINAL_MAX:
        return LOSBand.NOMINAL
    if flow_ped_m_min < BAND_BUILDING_MAX:
        return LOSBand.BUILDING
    return LOSBand.CRITICAL


def los_grade_for_flow(flow_ped_m_min: float) -> str:
    """Full Fruin grade A-F, for the operator console. The app never shows this."""
    for grade, upper in (
        ("A", LOS_A_MAX),
        ("B", LOS_B_MAX),
        ("C", LOS_C_MAX),
        ("D", LOS_D_MAX),
        ("E", LOS_E_MAX),
    ):
        if flow_ped_m_min < upper:
            return grade
    return "F"


# --------------------------------------------------------------------------
# Pedestrian movement.
# Source: standard pedestrian planning figures; Fruin and subsequent replication.
# --------------------------------------------------------------------------

FREE_FLOW_SPEED_MS = 1.34
"""Free-flow walking speed, m/s. Used only as a prior -- per-zone observed speed
supersedes it wherever enough samples exist (plan/standards.md section 3)."""

JAM_DENSITY_PERSONS_M2 = 4.0
"""Jam density: ~0.25 m^2 per person. Flow approaches zero here."""

OBSERVED_HIGH_DENSITY_PERSONS_M2 = 4.7
"""Reported around large stadium events. Sanity ceiling for density estimates."""


# --------------------------------------------------------------------------
# Values that must be MEASURED, never assumed.
#
# Listed so that a reviewer can grep for them. Any literal standing in for one of
# these is a defect, not a default.
# --------------------------------------------------------------------------

MEASURED_NOT_ASSUMED = (
    "participation_rate",     # unique nodes / attendance, or capture-recapture
    "zone_capacity",          # observed peak sustained flow
    "corridor_width",         # lateral spread of traces
    "walking_speed",          # per-zone observed distribution
    "time_to_congestion",     # fitted to observed onset-to-saturation
    "prediction_confidence",  # node count, freshness, accuracy, stability
    "radio_range",            # walk-tested peer range in a standing crowd
    "hop_latency",            # observed discovery-to-transfer time per hop
    "uplink_coverage",        # fraction of nodes within reach of an uplink
)


# --------------------------------------------------------------------------
# Density thresholds, DERIVED from the flow thresholds above.
#
# Fruin's LOS is stated as a flow rate, but flow is not monotonic in density:
# it rises, peaks at capacity, then collapses. A jammed corridor and an empty
# one both show low flow. So flow alone cannot classify a zone -- and worse, the
# LOS E/F boundary of 82 ped/m/min sits ABOVE the physical maximum of ~80.4,
# making a flow-defined CRITICAL band unreachable in principle.
#
# The fix is to classify on density, using the densities that produce Fruin's
# flow boundaries on the free-flow branch. Inverting the Greenshields relation
#
#     flow = 60 * v_free * d * (1 - d / jam)
#
# for a target flow F gives the quadratic
#
#     d^2 - jam*d + (F * jam) / (60 * v_free) = 0
#
# whose lower root is the free-flow-branch density. These are computed, not
# typed: change a source constant and the boundaries move with it.
# --------------------------------------------------------------------------

def density_for_flow(
    flow_ped_m_min: float,
    free_speed_ms: float = FREE_FLOW_SPEED_MS,
    jam_density: float = JAM_DENSITY_PERSONS_M2,
) -> float | None:
    """Free-flow-branch density that produces a given flow rate.

    Returns None if the flow is above the physical maximum — which is itself a
    useful answer, and how the 82 ped/m/min discrepancy was found.
    """
    import math

    c = (flow_ped_m_min * jam_density) / (60.0 * free_speed_ms)
    disc = jam_density * jam_density - 4.0 * c
    if disc < 0:
        return None
    return (jam_density - math.sqrt(disc)) / 2.0


CAPACITY_DENSITY = JAM_DENSITY_PERSONS_M2 / 2.0
"""Density at maximum flow. Beyond it, more people means LESS throughput —
the counter-intuitive fact the whole product rests on, and the reason a zone at
capacity is already critical however healthy its flow rate looks."""

DENSITY_NOMINAL_MAX = density_for_flow(BAND_NOMINAL_MAX) or 0.75
"""Density equivalent of the LOS C/D boundary, on the free-flow branch."""

DENSITY_BUILDING_MAX = CAPACITY_DENSITY
"""CRITICAL starts at capacity, not at a flow number.

Fruin's 82 ped/m/min is unreachable under this fundamental diagram (max 80.4),
so the boundary is taken at the flow maximum instead — which is the same event
the LOS E/F boundary was describing: the point at which flow stops improving and
begins to collapse."""


def band_for_density(persons_per_m2: float) -> LOSBand:
    """Classify by density. This is the authoritative classifier.

    `band_for_flow` remains valid for a zone known to be on the free-flow branch,
    but anything measured from real occupancy should use this one: it is
    single-valued, whereas flow is not.
    """
    if persons_per_m2 < DENSITY_NOMINAL_MAX:
        return LOSBand.NOMINAL
    if persons_per_m2 < DENSITY_BUILDING_MAX:
        return LOSBand.BUILDING
    return LOSBand.CRITICAL


# --------------------------------------------------------------------------
# Mesh transport — delay-tolerant routing parameters.
#
# Cited where the literature gives a number, ASSUMED_-prefixed where it is a
# deployment choice nobody has published. The prefix is greppable on purpose:
# every ASSUMED_ constant below is a claim this project is making, not one it
# is inheriting, and each carries its reasoning.
#
# Source: Lindgren, Doria & Schelen, "Probabilistic Routing in Intermittently
# Connected Networks" (SIGMOBILE MC2R 2003), standardised as RFC 6693.
# Source: Spyropoulos, Psounis & Raghavendra, "Spray and Wait" (WDTN 2005).
# Source: Andres, Bordenabe, Chatzikokolakis & Palamidessi,
#         "Geo-Indistinguishability" (ACM CCS 2013).
# --------------------------------------------------------------------------

ASSUMED_RADIO_RANGE_CROWD_M = 30.0
"""Usable peer-to-peer radio range in a dense standing crowd, metres.

BLE and Wi-Fi Aware are specified around 100 m line of sight. A crowd is not line
of sight: 2.4 GHz is absorbed by bodies, and the phone is in a pocket against
one. 30 m is ASSUMED as the working figure and is a simulator parameter
everywhere it appears, so a single measured walk-test replaces it."""

PROPHET_P_INIT = 0.75
"""Delivery-predictability increment on a direct encounter.

P(a,b) <- P(a,b) + (1 - P(a,b)) * P_INIT. Lindgren et al. section 3.1."""

PROPHET_BETA = 0.25
"""Transitivity scaling: how much of a peer's predictability rubs off.

P(a,c) <- P(a,c) + (1 - P(a,c)) * P(a,b) * P(b,c) * BETA. This term is why
routing toward *connectivity* needs no special case: with the uplink as the
destination, "how likely am I to meet someone who reaches the internet" is
computed by the same recursion as any other destination."""

PROPHET_GAMMA = 0.98
"""Aging factor per time unit: P <- P * GAMMA^k. A node met once an hour ago is
not the node you should hand a message to now."""


def prophet_time_unit_s(
    radio_range_m: float = ASSUMED_RADIO_RANGE_CROWD_M,
    walk_speed_ms: float = FREE_FLOW_SPEED_MS,
    gamma: float = PROPHET_GAMMA,
) -> float:
    """The 'time unit' GAMMA ages per. Derived, and it has to be.

    Lindgren et al. publish GAMMA but leave the unit it applies to unspecified,
    and RFC 6693 makes it configurable — so a literal here would be the one
    number in the protocol that came from nowhere. It is also the number that
    decides whether the protocol works at all: aging is what makes a stale
    acquaintance stop looking like a route, and if the half-life is long compared
    to the encounter rate then aging is a no-op, every node's predictability
    ratchets to 1, and the forwarding test degenerates into a coin flip on
    floating-point noise. Measured here: at a 17-minute half-life the median
    predictability was 0.99 and the interquartile range 0.01.

    The right timescale is how long an encounter stays evidence about who is
    nearby NOW, which is how long it takes a node's radio neighbourhood to turn
    over: range / walking speed. Setting the half-life to that and solving
    GAMMA^(t/unit) = 1/2 gives the unit.
    """
    import math

    half_life_s = radio_range_m / walk_speed_ms
    return half_life_s * math.log(1.0 / gamma) / math.log(2.0)


PROPHET_TIME_UNIT_S = prophet_time_unit_s()
"""~0.65 s at a 30 m range and 1.34 m/s, i.e. a 22-second half-life."""

ASSUMED_SPRAY_COPY_SCALING = 1.0
"""Constant of proportionality in L = k * sqrt(M) for Spray-and-Wait.

Spyropoulos et al. show the copy count needed to stay within a fixed factor of
optimal delay grows as sqrt(M) in the population M; they do not fix k for an
arbitrary deployment, so k is ASSUMED 1 and the resulting L is reported by the
simulator rather than trusted. Getting k wrong costs delay, not correctness."""


def spray_copies_for(reachable_nodes: int) -> int:
    """Spray-and-Wait copy bound L for a population of M reachable nodes.

    Computed, not typed: L = ceil(k * sqrt(M)), floored at 2 because binary
    spray with a single copy degenerates to direct delivery.
    """
    import math

    if reachable_nodes <= 0:
        return 2
    return max(2, math.ceil(ASSUMED_SPRAY_COPY_SCALING * math.sqrt(reachable_nodes)))


MESH_TTL_MAX = 8
"""Hops remaining on a fresh message. Fixed by the MeshMessage contract (ttl<=8).

The consequence is worth stating plainly rather than discovering later: at
ASSUMED_RADIO_RANGE_CROWD_M, eight hops is a few hundred metres, not a circuit. The mesh is
a local harvesting network that feeds whatever uplink is nearby — it is not a
venue-wide backhaul. That is why uplink coverage is a reported metric (a region
with no uplink in range is simply not being heard) instead of an assumption."""

ASSUMED_HOP_LATENCY_S = 5.0
"""Time for one store-carry-forward hop: discovery, connect, transfer, metres.

Seconds, not milliseconds — Wi-Fi Aware/BLE discovery dominates. Used only to
size the dedupe window, where over-estimating is safe (a larger cache) and
under-estimating is not (a message outlives its own dedupe entry and loops)."""


def dedupe_retention_s(ttl: int = MESH_TTL_MAX,
                       hop_latency_s: float = ASSUMED_HOP_LATENCY_S) -> float:
    """How long a (source, sequence) must stay remembered.

    A message can survive at most `ttl` hops, so it cannot reappear later than
    ttl * hop_latency after it was first seen. Remembering for exactly that long
    is the smallest window that cannot loop.
    """
    return ttl * hop_latency_s


ASSUMED_MESH_BUFFER_MESSAGES = 256
"""Per-node store-carry-forward buffer, in messages.

Buffer exhaustion is the failure mode that makes epidemic routing unaffordable,
so the buffer must be bounded for that failure to be observable at all. At a few
hundred bytes per message this is well under 100 KB, which is nothing for a phone
and still small enough that flooding overruns it — which is the point."""

ASSUMED_URGENT_RELAYS_PER_MIN = 30.0
"""Token-bucket refill rate for URGENT (rate-limited epidemic) relays.

Epidemic routing is affordable for URGENT precisely because URGENT is rare; the
rate limit is what enforces the 'rare'. One relay every two seconds bounds the
radio duty cycle a compromised or malfunctioning node can impose on its
neighbours, while still saturating a neighbourhood within seconds."""

ASSUMED_URGENT_BURST_RELAYS = 15
"""Token-bucket capacity: one neighbourhood's worth of relays in a single burst.

An alert must reach everyone standing nearby on the first encounter, not be
metered out one peer at a time. The bucket then has to refill before the node can
flood again."""

ASSUMED_UPLINK_BATTERY_RESERVE = 0.20
"""Battery fraction below which a node will not be elected uplink.

A phone that dies is not a node, and a spectator whose battery we drained will
uninstall before the next race. Election is lexicographic (see the mesh module)
precisely so this is a hard floor rather than a weight traded against throughput."""

GEOIND_PRIVACY_LEVEL = 1.3862943611198906  # math.log(4)
"""Privacy level l in the geo-indistinguishability definition: a mechanism is
l-private within radius r when eps = l / r. ln(4) is the level used in the
worked example of Andres et al. (CCS 2013) and is the one this system reports."""

ASSUMED_GEOIND_RADIUS_M = 50.0
"""Radius r within which a reported position is indistinguishable, metres.

Andres et al. illustrate with r = 200 m, which is city scale: at a circuit it
would place a fragment in the wrong grandstand entirely. 50 m is ASSUMED as the
venue-scale equivalent — roughly a grandstand block, so a fragment cannot say
which entrance one person used, while the aggregate map is built at zone scale
where the zero-mean noise averages out at 1/sqrt(n). This number is the whole
privacy/utility trade and should be argued about explicitly, which is why it is
one named constant rather than an epsilon sprinkled through the code."""

GEOIND_EPSILON_VENUE = GEOIND_PRIVACY_LEVEL / ASSUMED_GEOIND_RADIUS_M
"""eps = l / r, the geo-indistinguishability parameter actually applied on device.

Derived, never typed: move the radius and epsilon moves with it. Every
TraceFragment records the epsilon it was built with, because a privacy claim that
is not attached to the data it describes is not checkable."""

ASSUMED_FRAGMENT_MAX_DURATION_S = 120.0
"""Longest span a single TraceFragment may cover, seconds.

Geo-indistinguishability bounds what one *point* reveals; it does not bound what
a long sequence of correlated points reveals, and a full-day trace is
re-identifiable however noisy each point is. Two minutes is ASSUMED as short
enough to be a corridor observation and too short to be an itinerary."""
